### PR TITLE
CoinMetrics DC sample

### PR DESCRIPTION
**Pull request by denispon**
https://unboundtech.atlassian.net/browse/EKP-4668

### CODE REVIEW

File: data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java [hunk 1]
```diff
@@ -154,7 +155,12 @@ public static void calcRiskForSignRequests(){
 
             TransactionHandler handler = getTransactionHandler(signRequest);
 
-            List<DetailedTransaction> detailedTransactions = handler.decode(signRequest.dataToSign, signRequest.rawTransactions, signRequest. publicKeys, Collections.emptySet());
+            List<DetailedTransaction> detailedTransactions = null;
+            try {
+                detailedTransactions = handler.decode(signRequest.dataToSign, signRequest.rawTransactions, signRequest. publicKeys, Collections.emptySet());
+            } catch (BadTransactionException e) {
+                e.printStackTrace();
```
Comment (unboundalex) [inline data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java:162]: the exception  needs to written to log, not to console.
also on error, the code needs to return and wait for the next interation. It cannot continue parsing the TX.

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        public String PriceUSD;
+
+        public CoinMetricsResponse(){}
+
+        public String getAsset() {
+            return asset;
+        }
+
+        public ZonedDateTime getTime() {
+            return ZonedDateTime.parse(time, DateTimeFormatter.ISO_ZONED_DATE_TIME);
```
Comment (unboundalex) [inline data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java:90]: extract business logic from POJO.

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+    public static class CoinMetricsResponse {
+        public String asset;
+        public String time;
+        public String PriceUSD;
```
Comment (unboundalex) [inline data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java:81]: Jackson can parse directly to BigDecimal

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            }
+
+            Optional<TransactionHandler> handler = getTransactionHandler(signRequest);
+
+            if(handler.isPresent()){
+                List<DetailedTransaction> detailedTransactions = new ArrayList<>();
+                try {
+                    detailedTransactions = handler.get().decode(signRequest.dataToSign, signRequest.rawTransactions, signRequest.publicKeys, new HashSet<>(signRequest.publicKeys));
+                } catch (BadTransactionException e) {
+                    System.err.println("failed to decode BTC transaction");
```
Comment (unboundalex) [inline data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:164]: missing "return;"

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                        .reduce(new BigInteger(String.valueOf(0L)), BigInteger::add);
+
+
+                BigDecimal btcInUSDRate = new BigDecimal(String.valueOf(-1L)).negate();
+                try {
+
+                    String yesterday = ZonedDateTime.now().minusDays(1).format(DateTimeFormatter.ISO_DATE);
+                    btcInUSDRate = service.getUSDPriceForBTC(yesterday);
+                } catch (JsonProcessingException e) {
+                    System.err.println("failed to get btcInUSD rate" + e.getMessage());
```
Comment (unboundalex) [inline data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:178]: missing "return;"

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                }
+
+                if(btcInUSDRate.compareTo(BigDecimal.valueOf(0L)) == -1){
+                    System.err.println("failed to get btcInUSD rate");
```
Comment (unboundalex) [inline data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:182]: missing "return;"

File: data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java (hunk 1 above)
Comment (denispon) [reply data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java:162]: i think your code is not updated...can u try to pull it again

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java (hunk 2 above)
Comment (denispon) [reply data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/CoinMetricsQueryService.java]: done

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java (hunk 6 above)
Comment (denispon) [reply data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:182]: done

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java (hunk 5 above)
Comment (denispon) [reply data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:178]: done

File: data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java (hunk 4 above)
Comment (denispon) [reply data-collectors/coinmetrics/src/main/java/com/unboundtech/casp/datacollector/coinmetrics/sample/Main.java:164]: done

File: data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java (hunk 1 above)
Comment (denispon) [reply data-collectors/ciphertrace/src/main/java/com/unboundtech/casp/datacollector/cyphertrace/sample/Main.java:162]: done

**Comment by denispon** [comment]
@unboundalex ready for additional review

**Review by unboundalex — APPROVED** [review]

**Pull request merged**